Studio encoding profile performance improved

wsmirnow opened the pull request:
The studio-upload workflow uses an dynamic encoding profile for generating the delivery version of the videos. This profile is currently really slow. I've changed the x264 encoder preset from slower to faster. On the way the buffer size for each quality was fixed to hold 2 sec. of video.

### Your pull request should…

* [ ] have a concise title
* [ ] [close an accompanying issue](https://help.github.com/en/articles/closing-issues-using-keywords) if one exists
* [ ] be against the correct branch (features can only go into develop)
* [ ] include migration scripts and documentation, if appropriate
* [ ] pass automated testing
* [ ] have a clean commit history
* [ ] have proper commit messages (title and body) for all commits
* [ ] have appropriate tags applied

wsmirnow committed:
Speedup studio adaptive streaming optimized encoding profile

The studio-upload workflow uses an dynamic encoding profile for generating the delivery version of the videos. This profile is currently really slow. I've changed the x264 encoder preset from slower to faster. On the way the buffer size for each quality was fixed to hold 2 sec. of video.

lkiesow commented on etc/encoding/adaptive-streaming-movies.properties:
```diff
@@ -108,33 +108,33 @@ profile.studio.adaptive-parallel.http.suffix.360p-quality = -360p-quality.mp4
-    -c:v libx264 -preset slower -tune film -pix_fmt yuv420p \
-    -x264opts keyint=25:min-keyint=25:no-scenecut -crf 20 -maxrate 14800k -bufsize 18000k \
-    -c:a aac -b:a 196k -ac 2 \
+    -c:v libx264 preset faster -tune film -pix_fmt yuv420p \
```
You accidentally deleted the `-`:

```suggestion
    -c:v libx264 -preset faster -tune film -pix_fmt yuv420p \
```

lkiesow commented on etc/encoding/adaptive-streaming-movies.properties:
```diff
@@ -108,33 +108,33 @@ profile.studio.adaptive-parallel.http.suffix.360p-quality = -360p-quality.mp4
-    -x264opts keyint=25:min-keyint=25:no-scenecut -crf 20 -maxrate 14800k -bufsize 18000k \
-    -c:a aac -b:a 196k -ac 2 \
+    -c:v libx264 preset faster -tune film -pix_fmt yuv420p \
+    -x264opts keyint=25:min-keyint=25:no-scenecut -crf 23 -maxrate 14800k -bufsize 29600k \
```
`-crf 23` is the default value. We can just drop it.

```suggestion
    -x264opts keyint=25:min-keyint=25:no-scenecut -maxrate 14800k -bufsize 29600k \
```

lkiesow commented on etc/encoding/adaptive-streaming-movies.properties:
```diff
@@ -108,33 +108,33 @@ profile.studio.adaptive-parallel.http.suffix.360p-quality = -360p-quality.mp4
-    -c:a aac -b:a 196k -ac 2 \
+    -c:v libx264 preset faster -tune film -pix_fmt yuv420p \
+    -x264opts keyint=25:min-keyint=25:no-scenecut -crf 23 -maxrate 14800k -bufsize 29600k \
+    -c:a aac -b:a 128k -ac 2 \
```
:+1:  for `128k` audio for speech. Since we are only recording a microphone which is unlikely to be a stereo signal, should we go for mono? And if we go for mono, should we go for `96k`? So many questions :D

gregorydlogan replied:
How much are we *really* saving by dropping to mono?  I'd leave it at stereo, but I'm happy to see the drop to 128.

lkiesow replied:
The idea was that mono will give you better quality for the same fixed bandwith. Especially if the content is really mono anyway.

wsmirnow replied:
I agree with @gregorydlogan. Dropping to 128k make sense. I would prefer to keep stereo.

lkiesow reviewed (changes requested):
Not sure I fully agree with this. This will result in faster encodings but also larger file sizes and lower quality. You also additionally lowered the quality more by increasing the constant rate factor, although I think 23 is just fine for that. Looking at [expected encoding times](https://trac.ffmpeg.org/wiki/Encode/H.264#Howdothedifferentpresetsinfluenceencodingtime), could we settle in the middle and pick `-preset medium`? This should take only about half the time of `slower` but still produce smaller files.

> On the way the buffer size for each quality was fixed to hold 2 sec. of video.

Can you explain where the 2s come from?

wsmirnow committed:
Update etc/encoding/adaptive-streaming-movies.properties

typo fixed

Co-Authored-By: Lars Kiesow <lkiesow@uos.de>

wsmirnow committed:
studio encoding profile updated

removed -crf 23 ffmpeg option as it is the default value

gregorydlogan commented:
Any progress on this?

wsmirnow commented:
I'm just fine with current state. @lkiesow you too?

lkiesow commented:
My previous complaints have not been addressed:

> Not sure I fully agree with this. This will result in faster encodings but also larger file sizes and lower quality. You also additionally lowered the quality more by increasing the constant rate factor, although I think 23 is just fine for that. Looking at [expected encoding times](https://trac.ffmpeg.org/wiki/Encode/H.264#Howdothedifferentpresetsinfluenceencodingtime), could we settle in the middle and pick `-preset medium`? This should take only about half the time of `slower` but still produce smaller files.
> 
> > On the way the buffer size for each quality was fixed to hold 2 sec. of video.
> 
> Can you explain where the 2s come from?

wsmirnow committed:
studio workflow updated

The x264 preset is now set to medium, which is the default preset. For this reason the preset option was removed from studio encoding profile.

wsmirnow commented:
> My previous complaints have not been addressed:
> 
> > Not sure I fully agree with this. This will result in faster encodings but also larger file sizes and lower quality. You also additionally lowered the quality more by increasing the constant rate factor, although I think 23 is just fine for that. Looking at [expected encoding times](https://trac.ffmpeg.org/wiki/Encode/H.264#Howdothedifferentpresetsinfluenceencodingtime), could we settle in the middle and pick `-preset medium`? This should take only about half the time of `slower` but still produce smaller files.
> > > On the way the buffer size for each quality was fixed to hold 2 sec. of video.
> > 
> > 
> > Can you explain where the 2s come from?

I've updated the x264 preset (removed option and go with `medium`). In my opinion the buffer size value of 2 sec isn't high but also a bit more then 1 sec. Here we should be safe. I don't know any more, why I choose this value. If you have any suggestions, let me know.

lkiesow commented:
> Can you explain where the 2s come from?

What I'm actually wondering is what's the relation of `-bufsize 29600k` and `2s`. Where do the two seconds come from? Why is that two seconds?

wsmirnow commented:
[Here](https://trac.ffmpeg.org/wiki/Limiting%20the%20output%20bitrate) you can find some thoughts about `bufsize` and `max-/minrate`.

wsmirnow committed:
studio encoding profile updated

Set bufsize for each quality to maxrate value. For more information see the ffmpeg documentation at
https://trac.ffmpeg.org/wiki/Limiting%20the%20output%20bitrate

wsmirnow commented:
@gregorydlogan the suggestions are solved now and the PR is ready for review.
